crates/cubecl-core/src/codegen/metadata.rs

```diff
@@ -0,0 +1,144 @@
+//! Metadata helpers to easily get offsets etc.
+//!
+//! Conceptually, metadata is represented like this:
+//! ```rust
+//! struct Metadata<const NUM_BUFS: usize, const NUM_EXT: usize> {
+//!     base: BaseMeta<NUM_BUFS>,
+//!     extended: ExtendedMeta<NUM_EXT>,
+//! }
+//!
+//! struct BaseMeta<const N: usize> {
+//!     buffer_lengths: [u32; N],
+//!     logical_lengths: [u32; N],
+//! }
+//!
+//! struct ExtendedMeta<const N: usize> {
+//!     ranks: [u32; N],
+//!     shape_offsets: [usize; N],
+//!     stride_offsets: [usize; N],
+//!     shapes: Vec<u32>,
+//!     strides: Vec<u32>
+//! }
```

Comment:
Maybe we could have more precise names for each metadata block:

```rust
 struct Metadata<const NB: usize, const NT: usize> {
    buffer: BufferMetadata<NB>,
    tensor: TensorMetadata<NT>,
}
```

Reply:
I kind of wanted to decouple it from the specific `Tensor` name because we may have other types with extended metadata at some point. Offsets allow for dynamically sized metadata so you could theoretically make any kind of data structure that has additional metadata associated with it, as long as it could be encoded as series of u32. That's why I wanted to use the more generic `ExtendedMetadata`